### ISSUE TITLE
Cannonical pangeo binder?

### ISSUE THREAD

**Issue by rabernat** [pangeo-data/pangeo-docker-images#18]
I am following this closely, as it is quite related to my ongoing work on refactoring our gallery ([status update here](https://discourse.pangeo.io/t/pangeo-example-gallery/493/10)).

I am very enthusiastic about the idea of having a sort of standard cannonical binder which we can use for many distinct example repos. I think this is what @scottyhq is developing in https://github.com/scottyhq/pangeodev-binder. The benefits of this approach are:
- Much faster to iterate binder development if you don't have to keep rebuilding the environment
- We could configure our binders to preload this image on all nodes, greatly speeding up launch time

The challenge is whether we can agree on what packages should be included.

Is https://github.com/scottyhq/pangeodev-binder basically ready? Can we move it to the pangeo org? I want to move forward with some stuff and it would be good to have this piece of the puzzle fixed.

**Comment by scottyhq** [comment]
> I am following this closely, as it is quite related to my ongoing work on refactoring our gallery (status update here).
Fantastic! I think this will be great. 

> I think this is what @scottyhq is developing in https://github.com/scottyhq/pangeodev-binder. 

That repo is mainly just for me testing these new images out. But I do like the github template repository approach over cookie cutter. 

> The challenge is whether we can agree on what packages should be included.

This is a huge challenge. For speed, simplicity, and pre-pulling images we could **require** that all examples use pangeo/pangeo-notebook:DATE. I think a better approach is to move people towards specifying explicit lists of packages for example notebooks, that way even if the Docker image is rebuilt, the environment is identical. We can do this with the new pangeodev/base-image, see:
https://github.com/scottyhq/pangeo-binder-explicit/tree/master/binder

> Is https://github.com/scottyhq/pangeodev-binder basically ready? Can we move it to the pangeo org? I want to move forward with some stuff and it would be good to have this piece of the puzzle fixed.

Almost. Let me push a few more changes this week to this repo. We should probably also add more tests to ensure the base-image and image building works as expected. I was thinking of adding another github action to actually run repo2docker... but if i understand correctly that is sort of what binderbot is doing!

**Comment by rabernat** [comment]
> This is a huge challenge. For speed, simplicity, and pre-pulling images we could **require** that all examples use pangeo/pangeo-notebook:DATE

Thinking about this a bit more. I think we should have a binder repo that automatically creates a tag (of the same name) each time a new pangeo-notebook image is published. Then we can launch notebooks with paths like:

```
https://mybinder.org/v2/gh/pangeo-data/pangeo-notebook-binder/2020.03.11
```

> I think a better approach is to move people towards specifying explicit lists of packages for example notebooks, that way even if the Docker image is rebuilt, the environment is identical.

This sounds hard. We want to lower the barrier to creating binders that just work. If we want to go this route, perhaps some automated script that generates the environment?

> I was thinking of adding another github action to actually run repo2docker... but if i understand correctly that is sort of what binderbot is doing!

Binderbot is interacting with binder as if it were a user, via the binder API. So if the image needs to be built, it will be built. If you want to see what it's like to use binderbot within a workflow, check this out: https://github.com/pangeo-gallery/example-gallery/blob/b8a6cdfbeb2a511f8b3e9e9c9dd7db4c57750b99/.github/workflows/binderbot.yaml#L41-L47.

**Comment by rabernat** [comment]
Just a ping here. As I mentioned to @TomAugspurger at the last Pangeo call, I would really love to have an official pangeo binder to point to for my gallery work. Doesn’t have to be perfect, but perhaps we can get a repo started so I can start debugging.

**Comment by TomAugspurger** [comment]
@rabernat does a GitHub template repository like https://github.com/pangeo-data/pangeo-binder-template work, with the changes that the default "launch binder"  URL is configured to an nbgitpuller URL? As long as all your dependencies are in the pangeo-notebook environment already, you wont' need to build a docker image.

**Comment by TomAugspurger** [comment]
Hmm maybe nbgitpuller isn't an option for binder, since (IIUC) it requires the users to already be authenticated. Perhaps we can mess with the URL binder redirects you too when launching the hub...

**Comment by jhamman** [comment]
I think @scottyhq has been using nbgitpuller and binder together.

**Comment by rabernat** [comment]
I am definitely planning to lean on nbgitpuller to separate binder repo from notebook content. [This thread](https://discourse.jupyter.org/t/tip-embed-custom-github-content-in-a-binder-link-with-nbgitpuller/922/25) has become my bible.

But for that approach to work, I still need to know which repo to point to by default. That's why I'm hoping for a cannonical pangeo binder repo.

**Comment by TomAugspurger** [comment]
> I still need to know which repo to point to by default

What's a "repo" here? A repo with just a `binder/` folder that has the "canonical" pangeo packages? So then when you launch a binder, it'll find that image already built in its cache?

If so, then I think what we want is

**Case 1: no custom packages**

A user has some notebooks to share that uses just core packages (xarray, dask, ...). To share this in a binder, they just need a URL like `https://hub.pangeo.io/hub/user-redirect/git-pull?repo=https%3A%2F%2Fgithub.com%2FTomAugspurger%2Fpangeo-basic&urlpath=lab%2Ftree%2Fpangeo-basic%2F`. Breaking that apart

* Jupyterhub: `hub.pangeo.io`
* Repo: `TomAugspurger/pangeo-basic`. A (dummy) git repository with notebooks
* urlpath: use jupyterlab by default

I *think* all of that works, as long as `hub.pangeo.io` has nbgitpuller installed.

**Case 2: requires custom packages**

Then I think you're back to either

2a. Including your own `binder/environment.yaml`, so you're building from scratch
2b. Something like https://github.com/dask/distributed/pull/3216 (which installs packages on workers as they startup. Fragile, but maybe good enough)

I'll see if I can ensure that option 1 is working today.

**Comment by rabernat** [comment]
> What's a "repo" here? A repo with just a `binder/` folder that has the "canonical" pangeo packages?

Yes, plus whatever is need to make dask gateway work, dask dashboard, etc. It's more than just python packages.

> * Repo: `TomAugspurger/pangeo-basic`. A (dummy) git repository with notebooks

Yes, this is exactly right. But I would like to bless some repo as the "official" repo for a basic binder. Furthermore, it would be nice to have its tags sync with pangeo-stacks tags, so we can point to a specific release of pangeo-notebook.

> Case 2: requires custom packages

In this case, we can also put the binder in a separate branch of the *same* repo, to avoid constant rebuilding.

**Comment by TomAugspurger** [comment]
> Yes, plus whatever is need to make dask gateway work, dask dashboard, etc. It's more than just python packages.

Gotcha, thanks. IMO, the right place to specify that is in the `pangeo-notebook` metapackage. So IMO something like https://github.com/TomAugspurger/pangeo-binder-test is the right structure for a canonical pangeo binder repo

1. A `binder/environment.yaml` that includes `pangeo-notebook` and `nbgitpuller`
2. A postBuild that executes `gitpuller <repo with the content to share>`

Then the only thing the user sharing contents needs is the URL like `https://staging.binder.pangeo.io/v2/gh/pangeo-data/canonical-binder/master?urlpath=git-pull?repo=https://github.com/TomAugspurger/pangeo-basic`.

The first part `https://staging.binder.pangeo.io/v2/gh/pangeo-data/canonical-binder/` is the same for everyone sharing content. The only thing that changes is the `urlpath=git-pull?repo=https://github.com/TomAugspurger/pangeo-basic`.

> Furthermore, it would be nice to have its tags sync with pangeo-stacks tags, so we can point to a specific release of pangeo-notebook.

Yep, that's easy to do (and hopefully automate). Then for a reproducible build, the URL would be `https://staging.binder.pangeo.io/v2/gh/pangeo-binder/canonical-binder/2020.03.27`.

I'll have a test in a bit. If that works, all we'll need to do is make the `pangeo-binder/canonical-binder` repository.

**Comment by rabernat** [comment]
> I'll have a test in a bit. If that works, all we'll need to do is make the `pangeo-binder/canonical-binder` repository.

This might make sense to live in the `pangeo-gallery` org.

**Comment by rabernat** [comment]
You can see all of this in progress here: http://gallery.pangeo.io/repos/pangeo-gallery/example-gallery/01_example-notebook.html

**Comment by TomAugspurger** [comment]
@rabernat here's an example URL

https://staging.binder.pangeo.io/v2/gh/TomAugspurger/pangeo-binder-test/master?urlpath=git-pull?repo=https://github.com/pangeo-gallery/example-gallery


The only things to change

1. Remove staging
2. `v2/gh/TomAugspurger/pangeo-binder-test/` -> `pangeo-gallery/canonical-binder` (which has the same contents as https://github.com/TomAugspurger/pangeo-binder-test
3. Figure out jupyerlab by default.

**Comment by TomAugspurger** [comment]
@scottyhq have you used binder + nbgitpuller? Do you know what kind of URL is necessary to get dropped into jupyterlab? I thought that

https://staging.binder.pangeo.io/v2/gh/TomAugspurger/pangeo-binder-test/master?urlpath=git-pull?repo=https%3A%2F%2Fgithub.com%2FTomAugspurger%2Fpangeo-basic&urlpath=lab%2Ftree%2Fpangeo-basic%2F

would do it, but apparently not.

**Comment by rabernat** [comment]
> @scottyhq have you used binder + nbgitpuller? Do you know what kind of URL is necessary to get dropped into jupyterlab? 

http://mybinder.org/v2/gh/choldgraf/binder-sandbox/master/?urlpath=git-pull?repo=https://github.com/pangeo-gallery/example-gallery%26amp%3Burlpath=lab/tree/example-gallery

I've got this part figured out. What I really needed help with is the cannonical binder repo itself, not the linking semantics. That seems to be sorted out, so thanks!

**Comment by TomAugspurger** [comment]
> What I really needed help with is the cannonical binder repo itself

I think make a new repo (`pangeo-gallery/canonical-binder`) with the contents of https://github.com/TomAugspurger/pangeo-binder-test.

And if there are missing packages that you consider essential, they'd be added to https://github.com/conda-forge/pangeo-notebook-feedstock/blob/master/recipe/meta.yaml

**Comment by TomAugspurger** [comment]
@rabernat thanks for the invite. Do you want me to set up the pangeo-gallery/canonical-binder repository?

**Comment by rabernat** [comment]
> Do you want me to set up the pangeo-gallery/canonical-binder repository?

What name should we use? Maybe `default-binder` is better?

**Comment by TomAugspurger** [comment]
Agreed `default-binder` sounds better.

**Comment by scottyhq** [comment]
Thanks @TomAugspurger and @rabernat . I'm just catching up on this, and will check-in on the poets call tmrw at 10. I think the plan sounds good! 

An important aspect of this working is that the dask-gateway configuration is now done at the hub/binderhub definition level and not per-image. So we will need to require gallery notebooks to use dask-gateway instead of dask-kubernetes.
https://github.com/pangeo-data/pangeo-binder/blob/da809585b68d50134ea0fd545a220ba186caaedd/pangeo-binder/values.yaml#L190

We need to implement https://github.com/pangeo-data/pangeo-binder/issues/129 so that example notebooks can read from requester-pays buckets. @TomAugspurger - are you able to take a pass at that? 

One other thing to consider is using our new base-image in the binder folder. If you just use the pangeo-notebook metapackage, you will be running through the standard repo2docker conda-builder. https://github.com/scottyhq/pangeodev-binder/blob/master/binder/Dockerfile . I'll be tagging a new image today with some changes @tjcrone made over the weekend

**Comment by TomAugspurger** [comment]
> We need to implement pangeo-data/pangeo-binder#129 so that example notebooks can read from requester-pays buckets. @TomAugspurger - are you able to take a pass at that?

I'm happy to look at that, but it seems to me that it's independent from the issue here. I believe that https://github.com/pangeo-data/pangeo-example-notebooks/blob/master/cesm-pop-highres-ocean.ipynb is reading from a requester pays bucket, and that workers on the binder.pangeo.io binder.

> One other thing to consider is using our new base-image in the binder folder.

What all is necessary for that? Just a `binder/Dockerfile` with the line `FROM pangeo/base-image:2020.03.27`?

**Comment by scottyhq** [comment]
> I believe that https://github.com/pangeo-data/pangeo-example-notebooks/blob/master/cesm-pop-highres-ocean.ipynb is reading from a requester pays bucket, and that workers on the binder.pangeo.io binder.

I'm not sure what the bucket configurations are on GCE, but either you have to assign permissions to the bucket itself to allow reading (e.g. from GCE clusters), or what were doing on AWS is linking an IAM role to the kubernetes service accounts themselves. I think this is the way to go b/c you can have different permissions at the pod level. There's a rather lengthy discussion of this from a while back here https://github.com/pangeo-data/pangeo-cloud-federation/issues/485

> What all is necessary for that? Just a binder/Dockerfile with the line FROM pangeo/base-image:2020.03.27

Yes.

**Comment by scottyhq** [comment]
closed by #250, although you may still want the `default-binder` repository for further customization (workspace definitions, start scripts etc).
